### ISSUE TITLE
Segfault in 4.54 on aarch64

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#31101]
Workflow 4.54 step 3 segfaults in CMSSW_11_2 2020-08-07-1100 with
```
Thread 4 (Thread 0xffff19988460 (LWP 151054)):
#0  0x0000ffff7c9aa9c4 in nanosleep () from /lib64/libc.so.6
#1  0x0000ffff7c9aa678 in sleep () from /lib64/libc.so.6
#2  0x0000ffff781643b8 in sig_pause_for_stacktrace () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginFWCoreServicesPlugins.so
#3  &lt;signal handler called&gt;
#4  operator delete (ptr=0xfffe8d664000) at src/jemalloc_cpp.cpp:107
#5  0x0000ffff0df489ac in ConvBremPFTrackFinder::runConvBremFinder(edm::Handle&lt;std::vector&lt;reco::PFRecTrack, std::allocator&lt;reco::PFRecTrack&gt; &gt; &gt; const&amp, edm::Handle&lt;std::vector&lt;reco::Vertex, std::allocator&lt;reco::Vertex&gt; &gt; &gt; const&amp, edm::Handle&lt;std::vector&lt;reco::PFDisplacedTrackerVertex, std::allocator&lt;reco::PFDisplacedTrackerVertex&gt; &gt; &gt; const&amp, edm::Handle&lt;std::vector&lt;reco::PFConversion, std::allocator&lt;reco::PFConversion&gt; &gt; &gt; const&amp, edm::Handle&lt;std::vector&lt;reco::PFV0, std::allocator&lt;reco::PFV0&gt; &gt; &gt; const&amp, convbremhelpers::HeavyObjectCache const*, bool, bool, bool, std::vector&lt;reco::PFCluster, std::allocator&lt;reco::PFCluster&gt; &gt; const&amp, reco::GsfPFRecTrack const&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libRecoParticleFlowPFTracking.so
#6  0x0000ffff03c732c4 in PFElecTkProducer::produce(edm::Event&amp, edm::EventSetup const&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginRecoParticleFlowPFTrackingPlugins.so
#7  0x0000ffff7eb69ea8 in edm::stream::EDProducerAdaptorBase::doEvent(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ActivityRegistry*, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#8  0x0000ffff7eb45db0 in edm::WorkerT&lt;edm::stream::EDProducerAdaptorBase&gt;::implDo(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so

Thread 3 (Thread 0xffff1a398460 (LWP 151053)):
#0  0x0000ffff7c9aa9c4 in nanosleep () from /lib64/libc.so.6
#1  0x0000ffff7c9aa678 in sleep () from /lib64/libc.so.6
#2  0x0000ffff781643b8 in sig_pause_for_stacktrace () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginFWCoreServicesPlugins.so
#3  &lt;signal handler called&gt;
#4  0x0000ffff07a314cc in ROOT::Math::SMatrix&lt;double, 17u, 17u, ROOT::Math::MatRepSym&lt;double, 17u&gt; &gt;::value_type ROOT::Math::meta_matrix_dot&lt;16u&gt;::g&lt;ROOT::Math::SMatrix&lt;double, 17u, 17u, ROOT::Math::MatRepSym&lt;double, 17u&gt; &gt;, ROOT::Math::SMatrix&lt;double, 17u, 17u, ROOT::Math::MatRepSym&lt;double, 17u&gt; &gt; &gt;(ROOT::Math::SMatrix&lt;double, 17u, 17u, ROOT::Math::MatRepSym&lt;double, 17u&gt; &gt; const&amp, ROOT::Math::SMatrix&lt;double, 17u, 17u, ROOT::Math::MatRepSym&lt;double, 17u&gt; &gt; const&amp, unsigned int, unsigned int) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libRecoEgammaEgammaPhotonAlgos.so
#5  0x0000ffff07a32d80 in KinematicConstrainedVertexUpdatorT&lt;2, 2&gt;::update(ROOT::Math::SVector&lt;double, 17u&gt; const&amp, ROOT::Math::SMatrix&lt;double, 17u, 17u, ROOT::Math::MatRepSym&lt;double, 17u&gt; &gt;&amp, std::vector&lt;KinematicState, std::allocator&lt;KinematicState&gt; &gt;&amp, Point3DBase&lt;float, GlobalTag&gt; const&amp, Vector3DBase&lt;float, GlobalTag&gt; const&amp, MultiTrackKinematicConstraintT&lt;2, 2&gt;*) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libRecoEgammaEgammaPhotonAlgos.so
#6  0x0000ffff07a33d04 in KinematicConstrainedVertexFitterT&lt;2, 2&gt;::fit(std::vector&lt;ReferenceCountingPointer&lt;KinematicParticle&gt;, std::allocator&lt;ReferenceCountingPointer&lt;KinematicParticle&gt; &gt; &gt; const&amp, MultiTrackKinematicConstraintT&lt;2, 2&gt;*, Point3DBase&lt;float, GlobalTag&gt;*) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libRecoEgammaEgammaPhotonAlgos.so
#7  0x0000ffff07a2a000 in ConversionVertexFinder::run(std::vector&lt;reco::TransientTrack, std::allocator&lt;reco::TransientTrack&gt; &gt; const&amp, reco::Vertex&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libRecoEgammaEgammaPhotonAlgos.so
#8  0x0000ffff07afb4d0 in ConversionProducer::checkVertex(reco::TransientTrack const&amp, reco::TransientTrack const&amp, MagneticField const*, reco::Vertex&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginRecoEgammaEgammaPhotonProducers.so
#9  0x0000ffff07afe92c in ConversionProducer::buildCollection(edm::Event&amp, edm::EventSetup const&amp, std::multimap&lt;float, edm::Ptr&lt;reco::ConversionTrack&gt;, std::less&lt;float&gt;, std::allocator&lt;std::pair&lt;float const, edm::Ptr&lt;reco::ConversionTrack&gt; &gt; &gt; &gt; const&amp, std::multimap&lt;double, edm::Ptr&lt;reco::CaloCluster&gt;, std::less&lt;double&gt;, std::allocator&lt;std::pair&lt;double const, edm::Ptr&lt;reco::CaloCluster&gt; &gt; &gt; &gt; const&amp, std::multimap&lt;double, edm::Ptr&lt;reco::CaloCluster&gt;, std::less&lt;double&gt;, std::allocator&lt;std::pair&lt;double const, edm::Ptr&lt;reco::CaloCluster&gt; &gt; &gt; &gt; const&amp, reco::Vertex const&amp, std::vector&lt;reco::Conversion, std::allocator&lt;reco::Conversion&gt; &gt;&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginRecoEgammaEgammaPhotonProducers.so
#10 0x0000ffff07b00908 in ConversionProducer::produce(edm::Event&amp, edm::EventSetup const&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginRecoEgammaEgammaPhotonProducers.so
#11 0x0000ffff7eb69ea8 in edm::stream::EDProducerAdaptorBase::doEvent(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ActivityRegistry*, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#12 0x0000ffff7eb45db0 in edm::WorkerT&lt;edm::stream::EDProducerAdaptorBase&gt;::implDo(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so

Thread 1 (Thread 0xffff7c220000 (LWP 148741)):
#0  0x0000ffff7c9aa9c4 in nanosleep () from /lib64/libc.so.6
#1  0x0000ffff7c9aa678 in sleep () from /lib64/libc.so.6
#2  0x0000ffff781643b8 in sig_pause_for_stacktrace () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginFWCoreServicesPlugins.so
#3  &lt;signal handler called&gt;
#4  0x0000ffff76a88474 in dqm::impl::MonitorElement::accessMut() () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libDQMServicesCore.so
#5  0x0000ffff76a83b18 in dqm::impl::MonitorElement::setBinContent(int, int, double) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libDQMServicesCore.so
#6  0x0000ffff0d452748 in tadqm::TrackAnalyzer::fillHistosForState(edm::EventSetup const&amp, reco::Track const&amp, std::__cxx11::basic_string&lt;char, std::char_traits&lt;char&gt;, std::allocator&lt;char&gt; &gt;) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginDQMTrackingMonitor.so
#7  0x0000ffff0d4554b8 in tadqm::TrackAnalyzer::analyze(edm::Event const&amp, edm::EventSetup const&amp, reco::Track const&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginDQMTrackingMonitor.so
#8  0x0000ffff0d49522c in TrackingMonitor::analyze(edm::Event const&amp, edm::EventSetup const&amp) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/pluginDQMTrackingMonitor.so
#9  0x0000ffff7eb69ea8 in edm::stream::EDProducerAdaptorBase::doEvent(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ActivityRegistry*, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so
#10 0x0000ffff7eb45db0 in edm::WorkerT&lt;edm::stream::EDProducerAdaptorBase&gt;::implDo(edm::EventPrincipal const&amp, edm::EventSetupImpl const&amp, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02640/slc7_aarch64_gcc820/cms/cmssw/CMSSW_11_2_X_2020-08-07-1100/lib/slc7_aarch64_gcc820/libFWCoreFramework.so

Current Modules:

Module: PFProducer:particleFlowTmp (crashed)
Module: TrackingMonitor:TrackerCollisionSelectedTrackMonCommongeneralTracks
Module: none
Module: none
Module: ConversionProducer:allConversions
```
Full log http://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_aarch64_gcc820/CMSSW_11_2_X_2020-08-07-1100/pyRelValMatrixLogs/run/4.54_RunEl2012B+RunEl2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT/step3_RunEl2012B+RunEl2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT.log

**Comment by makortel** [comment]
assign core

**Comment by makortel** [comment]
I'm puzzled with a few things in this crash

The job is configured to use 4 threads and 4 streams
```
%MSG-i ThreadStreamSetup:  (NoModuleName) 07-Aug-2020 18:34:15 CEST pre-events
setting # threads 4
setting # streams 4
%MSG
```
but the stack trace shows 5 `Module:` rows, and total 6 threads with TBB (in addition to the ones mentioned in the description, threads 9-11 are in `tbb::internal::futex_wait`).

The end of the strace says `PFProducer:particleFlowTmp (crashed)`, but there is no `PFProducer` elsewhere in the trace.

**Comment by makortel** [comment]
The line
```
#4  operator delete (ptr=0xfffe8d664000) at src/jemalloc_cpp.cpp:107
```
hints towards a memory problem _somewhere_ (not necessarily in `ConvBremPFTrackFinder`).

**Comment by Dr15Jones** [comment]
@makortel 
> The job is configured to use 4 threads and 4 streams
...
> but the stack trace shows 5 Module: rows, and total 6 threads with TBB (in addition to the ones mentioned in the description, threads 9-11 are in tbb::internal::futex_wait).

This is a consequence of having the additional `tbb::task::wait` needed to handle the prefetching of the EventSetup modules. tbb will sometimes decide to startup more threads. However, as far as I've been able to tell, it never concurrently uses more threads than the number we've specified in the configuration. Once we move to using all `esConsumes` that extra wait can be removed.
